File: sdks/typescript/src/uniswap/swap.ts
/**
 * `swap()` — agent-side execution after SBO3L's policy gate has cleared.
 *
 * Mock mode (default): returns a deterministic pseudo-tx-hash derived from
 * the SwapParams + a nonce. Lets the demo path run in CI without secrets.
 *
 * Live mode (`SBO3L_LIVE_ETH=1` + `SBO3L_ETH_RPC_URL` + `SBO3L_ETH_PRIVATE_KEY`
 * in env, OR explicit `env: SwapEnv` arg): constructs the SwapRouter02
 * `exactInputSingle` calldata, signs with the caller's private key, broadcasts
 * via `eth_sendRawTransaction`. Returns the resulting tx hash + Etherscan URL.
 *
 * The signing primitives (RLP-encoded EIP-1559 tx, secp256k1 ECDSA) live in
 * a vendored implementation here so the SDK has zero crypto deps. Add `viem`
 * if you need EIP-712 / Permit2 / multi-call paths — out of scope for v1.
 */

import { createHash } from "node:crypto";
import {
  EXACT_INPUT_SINGLE_SELECTOR,
  SEPOLIA_CHAIN_ID,
  SEPOLIA_SWAP_ROUTER_02,
  sepoliaEtherscanTxUrl,
  SEPOLIA_USDC,
  SEPOLIA_WETH,
} from "./sepolia.js";

/** Single-pool exact-input swap parameters. Mirrors Rust's `SwapParams`. */
export interface SwapParams {
  /** EIP-55 address of the token being sold (e.g. WETH or USDC). */
  tokenIn: string;
  /** EIP-55 address of the token being bought. */
  tokenOut: string;
  /** Pool fee tier in hundredths of a bip. 500 / 3000 / 10000. */
  fee: number;
  /** Recipient EOA — receives `tokenOut`. Usually the caller's address. */
  recipient: string;
  /** Exact amount of `tokenIn` to spend, in token's smallest unit (wei / micros). */
  amountIn: bigint;
  /** Slippage floor — derive from a recent quote, e.g. `quote * 99n / 100n`. */
  amountOutMinimum: bigint;
  /** Price ceiling. `0n` disables (most common). */
  sqrtPriceLimitX96?: bigint;
}

export interface SwapEnv {
  /** RPC URL (e.g. an Alchemy Sepolia URL). */
  rpcUrl: string;
  /** 32-byte private key as 0x-prefixed hex. */
  privateKeyHex: string;
  /** Sepolia chain id. Defaults to `SEPOLIA_CHAIN_ID`. */
  chainId?: number;
  /** SwapRouter02 address. Defaults to `SEPOLIA_SWAP_ROUTER_02`. */
  routerAddress?: string;
}

export interface SwapResult {
  /** `mock` when no live env is provided; `live` when on-chain tx broadcast succeeded. */
  mode: "mock" | "live";
  /** 0x-prefixed transaction hash. Pseudo-hash in mock mode (deterministic). */
  txHash: string;
  /** Sepolia Etherscan URL for the tx. */
  etherscanUrl: string;
  /** Hex-encoded calldata that was (or would have been) sent. */
  calldata: string;
  /** Router address used. */
  to: string;
}

/** Encode the SwapRouter02 `exactInputSingle` calldata. Mirrors Rust's `encode_exact_input_single`. */
export function encodeExactInputSingle(params: SwapParams): string {
  const tokenIn = parseAddress(params.tokenIn);
  const tokenOut = parseAddress(params.tokenOut);
  const recipient = parseAddress(params.recipient);
  const sqrtPrice = params.sqrtPriceLimitX96 ?? 0n;

  const out = new Uint8Array(4 + 32 * 7);
  // selector
  out.set(hexToBytes(EXACT_INPUT_SINGLE_SELECTOR), 0);
  // tokenIn
  out.set(addressPadded(tokenIn), 4);
  // tokenOut
  out.set(addressPadded(tokenOut), 4 + 32);
  // fee (uint24 → uint256)
  out.set(uintPadded(BigInt(params.fee)), 4 + 64);
  // recipient
  out.set(addressPadded(recipient), 4 + 96);
  // amountIn
  out.set(uintPadded(params.amountIn), 4 + 128);
  // amountOutMinimum
  out.set(uintPadded(params.amountOutMinimum), 4 + 160);
  // sqrtPriceLimitX96
  out.set(uintPadded(sqrtPrice), 4 + 192);

  return bytesToHex(out);
}

/**
 * Build + (in live mode) sign + broadcast a SwapRouter02 swap.
 *
 * In **mock mode**, returns a deterministic pseudo-tx-hash derived from
 * the calldata so different swap params produce different hashes. The
 * Etherscan URL is still real — judges can paste it later once the swap
 * actually fires in live mode.
 */
export async function swap(params: SwapParams, env?: SwapEnv): Promise<SwapResult> {
  const calldata = encodeExactInputSingle(params);
  const router = env?.routerAddress ?? SEPOLIA_SWAP_ROUTER_02;

  const liveEnabled = env !== undefined || process.env["SBO3L_LIVE_ETH"] === "1";
  if (!liveEnabled) {
    // Deterministic pseudo-hash so the same params always print the same
    // hash — useful for snapshot tests and demo reproducibility.
    const txHash = "0x" + sha256Hex(calldata + router).slice(0, 64);
    return {
      mode: "mock",
      txHash,
      etherscanUrl: sepoliaEtherscanTxUrl(txHash),
      calldata,
      to: router,
    };
  }

  const liveEnv = env ?? envFromProcess();
  const chainId = liveEnv.chainId ?? SEPOLIA_CHAIN_ID;
  const txHash = await broadcastEip1559(
    liveEnv.rpcUrl,
    liveEnv.privateKeyHex,
    chainId,
    router,
    calldata,
  );
  return {
    mode: "live",
    txHash,
    etherscanUrl: sepoliaEtherscanTxUrl(txHash),
    calldata,
    to: router,
  };
}

/* -------------------------------------------------------------------------- */
/*  Helpers                                                                    */
/* -------------------------------------------------------------------------- */

function envFromProcess(): SwapEnv {
  const rpcUrl = process.env["SBO3L_ETH_RPC_URL"];
  const privateKeyHex = process.env["SBO3L_ETH_PRIVATE_KEY"];
  if (rpcUrl === undefined || privateKeyHex === undefined) {
    throw new Error(
      "live swap requires SBO3L_ETH_RPC_URL + SBO3L_ETH_PRIVATE_KEY (or pass `env`)",
    );
  }
  return { rpcUrl, privateKeyHex };
}

function parseAddress(s: string): Uint8Array {
  const trimmed = s.startsWith("0x") || s.startsWith("0X") ? s.slice(2) : s;
  if (trimmed.length !== 40) {
    throw new Error(`address must be 0x + 40 hex chars, got ${trimmed.length}`);
  }
  return hexToBytes(trimmed);
}

const HEX_RE = /^[0-9a-fA-F]*$/;

function hexToBytes(hex: string): Uint8Array {
  const trimmed = hex.startsWith("0x") || hex.startsWith("0X") ? hex.slice(2) : hex;
  if (trimmed.length % 2 !== 0) throw new Error("hex must be even-length");
  // parseInt("gg", 16) returns NaN which silently coerces to 0 in `Uint8Array.set`,
  // letting malformed calldata reach the chain. Validate explicitly first.
  if (!HEX_RE.test(trimmed)) {
    throw new Error(`hex contains non-hex characters: ${JSON.stringify(hex)}`);
  }
  const out = new Uint8Array(trimmed.length / 2);
  for (let i = 0; i < out.length; i++) {
    out[i] = parseInt(trimmed.slice(i * 2, i * 2 + 2), 16);
  }
  return out;
}

function bytesToHex(bytes: Uint8Array): string {
  let s = "0x";
  for (const b of bytes) s += b.toString(16).padStart(2, "0");
  return s;
}

function addressPadded(addr: Uint8Array): Uint8Array {
  const out = new Uint8Array(32);
  out.set(addr, 12);
  return out;
}

function uintPadded(v: bigint): Uint8Array {
  if (v < 0n) throw new Error("uint must be non-negative");
  const out = new Uint8Array(32);
  let i = 31;
  let x = v;
  while (x > 0n && i >= 0) {
    out[i--] = Number(x & 0xffn);
    x >>= 8n;
  }
  return out;
}

function sha256Hex(input: string): string {
  return createHash("sha256").update(input).digest("hex");
}

/**
 * Sign + broadcast an EIP-1559 transaction via `viem` (optional peer dep).
 *
 * `viem` is dynamically imported so the SDK's *non-live* surface (calldata
 * encoding, mock-mode swap, demo smoke) keeps zero runtime dependencies.
 * Live mode requires consumers to install `viem` themselves —
 * `peerDependenciesMeta.viem.optional: true` flags it as non-mandatory at
 * install time so npm doesn't warn for the 99% of users who never enable
 * live mode.
 *
 * Why not vendor secp256k1+RLP: ~600 LoC of audited crypto would dwarf the
 * SDK's ~1500 LoC core, and viem's tree-shaken footprint (~30kB gzipped)
 * is smaller than what we'd ship by vendoring noble-secp256k1 ourselves.
 */
async function broadcastEip1559(
  rpcUrl: string,
  privateKeyHex: string,
  chainId: number,
  to: string,
  data: string,
): Promise<string> {
  const viem = await loadViem();
  const account = viem.privateKeyToAccount(normalisePrivateKey(privateKeyHex));
  const chain = {
    id: chainId,
    name: chainId === SEPOLIA_CHAIN_ID ? "sepolia" : `chain-${chainId}`,
    nativeCurrency: { name: "Ether", symbol: "ETH", decimals: 18 },
    rpcUrls: { default: { http: [rpcUrl] }, public: { http: [rpcUrl] } },
  } as const;
  const wallet = viem.createWalletClient({
    account,
    chain,
    transport: viem.http(rpcUrl),
  });

  // viem >= 1.x defaults sendTransaction to EIP-1559 (type 2). It pulls
  // maxFeePerGas / maxPriorityFeePerGas from the chain's fee oracle and
  // estimates gas. Result is the 0x-prefixed tx hash — the caller then
  // tracks inclusion via Etherscan or a public client.
  const hash = await wallet.sendTransaction({
    to: viem.getAddress(to),
    data: data as `0x${string}`,
    value: 0n,
  });
  return hash;
}

interface ViemModule {
  createWalletClient: (config: {
    account: unknown;
    chain: unknown;
    transport: unknown;
  }) => {
    sendTransaction: (args: {
      to: `0x${string}`;
      data: `0x${string}`;
      value: bigint;
    }) => Promise<`0x${string}`>;
  };
  http: (url: string) => unknown;
  privateKeyToAccount: (key: `0x${string}`) => unknown;
  getAddress: (a: string) => `0x${string}`;
}

async function loadViem(): Promise<ViemModule> {
  try {
    // Indirect through string variables so tsc doesn't try to statically
    // resolve `viem` at compile time. The SDK's `viem` peer dep is
    // intentionally optional: most consumers never enable live mode and
    // shouldn't need to install it just to typecheck.
    const viemSpec: string = "viem";
    const accountsSpec: string = "viem/accounts";
    const mod = (await import(/* @vite-ignore */ viemSpec)) as Partial<ViemModule>;
    const accountsMod = (await import(/* @vite-ignore */ accountsSpec)) as Partial<ViemModule>;
    const merged: Partial<ViemModule> = { ...mod, ...accountsMod };
    if (
      merged.createWalletClient === undefined ||
      merged.http === undefined ||
      merged.privateKeyToAccount === undefined ||
      merged.getAddress === undefined
    ) {
      throw new Error("viem present but missing expected exports");
    }
    return merged as ViemModule;
  } catch (err) {
    const cause = err instanceof Error ? err.message : String(err);
    throw new Error(
      "live swap requires the `viem` peer dependency. " +
        "Install with `npm i viem` (or `pnpm add viem` / `yarn add viem`). " +
        `Original load error: ${cause}`,
    );
  }
}

function normalisePrivateKey(raw: string): `0x${string}` {
  const trimmed = raw.startsWith("0x") || raw.startsWith("0X") ? raw.slice(2) : raw;
  if (trimmed.length !== 64 || !HEX_RE.test(trimmed)) {
    throw new Error("private key must be 32 bytes (64 hex chars), 0x prefix optional");
  }
  return ("0x" + trimmed.toLowerCase()) as `0x${string}`;
}

/** Re-export Sepolia constants so callers don't need a separate import. */
export { SEPOLIA_USDC, SEPOLIA_WETH };

File: sdks/python/sbo3l_sdk/uniswap/swap.py
"""`swap()` — agent-side execution after SBO3L's policy gate has cleared.

Mock mode (default): deterministic pseudo-tx-hash from SwapParams + router.
Live mode (`SBO3L_LIVE_ETH=1` env or `env: SwapEnv` arg): builds calldata
and signs/broadcasts via `web3.py`. Live mode requires `web3` extra:

    pip install "sbo3l-sdk[web3]"

The signing primitive is `web3.py`'s `Account.sign_transaction` (so the
private key never leaves the agent process). SBO3L's daemon is policy-only
— the no-key boundary is preserved.

The signing import is deferred so callers without web3 installed can still
use `encode_exact_input_single` + `swap()` in mock mode.
"""

from __future__ import annotations

import hashlib
import os
from dataclasses import dataclass
from typing import Any, Literal

from .sepolia import (
    EXACT_INPUT_SINGLE_SELECTOR,
    SEPOLIA_CHAIN_ID,
    SEPOLIA_SWAP_ROUTER_02,
    sepolia_etherscan_tx_url,
)


@dataclass(frozen=True, slots=True)
class SwapParams:
    """Single-pool exact-input swap parameters. Mirrors Rust's `SwapParams`
    and TS's `SwapParams`. All amounts in token's smallest unit."""

    token_in: str
    token_out: str
    fee: int
    recipient: str
    amount_in: int
    amount_out_minimum: int
    sqrt_price_limit_x96: int = 0


@dataclass(frozen=True, slots=True)
class SwapEnv:
    """Live-mode env overrides. Pass to `swap(env=...)` to bypass `os.environ`."""

    rpc_url: str
    private_key_hex: str
    chain_id: int = SEPOLIA_CHAIN_ID
    router_address: str = SEPOLIA_SWAP_ROUTER_02


@dataclass(frozen=True, slots=True)
class SwapResult:
    """`mock` mode: pseudo-tx-hash; `live` mode: real on-chain tx hash."""

    mode: Literal["mock", "live"]
    tx_hash: str
    etherscan_url: str
    calldata: str  # 0x-prefixed
    to: str  # router address


def encode_exact_input_single(p: SwapParams) -> str:
    """Encode the SwapRouter02 `exactInputSingle` calldata. Mirrors Rust's
    `encode_exact_input_single` and TS's `encodeExactInputSingle`. Returns
    a 0x-prefixed lowercase hex string."""

    out = bytearray()
    out += EXACT_INPUT_SINGLE_SELECTOR
    out += _address_padded(_parse_address(p.token_in))
    out += _address_padded(_parse_address(p.token_out))
    out += _uint_padded(p.fee)
    out += _address_padded(_parse_address(p.recipient))
    out += _uint_padded(p.amount_in)
    out += _uint_padded(p.amount_out_minimum)
    out += _uint_padded(p.sqrt_price_limit_x96)
    return "0x" + out.hex()


def swap(params: SwapParams, env: SwapEnv | None = None) -> SwapResult:
    """Build + (in live mode) sign + broadcast a SwapRouter02 swap.

    Mock mode (no `env` and `SBO3L_LIVE_ETH != "1"`): returns deterministic
    pseudo-tx-hash so demos run in CI without secrets.

    Live mode: requires `web3.py` installed. Raises `RuntimeError` with a
    clear hint if web3 isn't available.
    """

    calldata = encode_exact_input_single(params)
    router = env.router_address if env else SEPOLIA_SWAP_ROUTER_02

    live_enabled = env is not None or os.environ.get("SBO3L_LIVE_ETH") == "1"
    if not live_enabled:
        # Deterministic pseudo-hash matches TS implementation: sha256(calldata + router)[:64]
        h = hashlib.sha256((calldata + router).encode("ascii")).hexdigest()[:64]
        tx_hash = "0x" + h
        return SwapResult(
            mode="mock",
            tx_hash=tx_hash,
            etherscan_url=sepolia_etherscan_tx_url(tx_hash),
            calldata=calldata,
            to=router,
        )

    live_env = env if env else _env_from_process()
    tx_hash = _broadcast_legacy_tx(
        live_env.rpc_url,
        live_env.private_key_hex,
        live_env.chain_id,
        router,
        calldata,
    )
    return SwapResult(
        mode="live",
        tx_hash=tx_hash,
        etherscan_url=sepolia_etherscan_tx_url(tx_hash),
        calldata=calldata,
        to=router,
    )


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _env_from_process() -> SwapEnv:
    rpc_url = os.environ.get("SBO3L_ETH_RPC_URL")
    private_key_hex = os.environ.get("SBO3L_ETH_PRIVATE_KEY")
    if rpc_url is None or private_key_hex is None:
        raise RuntimeError(
            "live swap requires SBO3L_ETH_RPC_URL + SBO3L_ETH_PRIVATE_KEY (or pass `env`)"
        )
    return SwapEnv(rpc_url=rpc_url, private_key_hex=private_key_hex)


def _parse_address(s: str) -> bytes:
    trimmed = s[2:] if s.startswith(("0x", "0X")) else s
    if len(trimmed) != 40:
        raise ValueError(f"address must be 0x + 40 hex chars, got {len(trimmed)}")
    try:
        return bytes.fromhex(trimmed)
    except ValueError as e:
        raise ValueError(f"address contains non-hex character: {e}") from e


def _address_padded(addr: bytes) -> bytes:
    """Left-pad a 20-byte address to a 32-byte word."""

    if len(addr) != 20:
        raise ValueError("address must be exactly 20 bytes")
    return b"\x00" * 12 + addr


def _uint_padded(v: int) -> bytes:
    """Right-pack a non-negative int into a 32-byte big-endian word."""

    if v < 0:
        raise ValueError("uint must be non-negative")
    return v.to_bytes(32, "big")


def _broadcast_legacy_tx(
    rpc_url: str,
    private_key_hex: str,
    chain_id: int,
    to: str,
    data: str,
) -> str:
    """Sign + broadcast a legacy (type-0) transaction via web3.py.

    Sepolia accepts both legacy and EIP-1559; legacy is simpler (no fee
    oracle dependency) and sufficient for the demo path. Production
    callers should use EIP-1559 with a proper fee bumping strategy.
    """

    try:
        from web3 import Web3  # type: ignore[import-not-found]
    except ImportError as e:
        raise RuntimeError(
            'live swap requires `web3.py` — install with `pip install "sbo3l-sdk[web3]"` '
            "or `pip install web3`."
        ) from e

    w3 = Web3(Web3.HTTPProvider(rpc_url))
    if not w3.is_connected():
        raise RuntimeError(f"could not connect to RPC at {rpc_url}")

    account = w3.eth.account.from_key(private_key_hex)
    nonce = w3.eth.get_transaction_count(account.address, "pending")
    gas_price = w3.eth.gas_price

    tx: dict[str, Any] = {
        "nonce": nonce,
        "gasPrice": gas_price,
        "gas": 300_000,
        "to": Web3.to_checksum_address(to),
        "value": 0,
        "data": data,
        "chainId": chain_id,
    }

    signed = account.sign_transaction(tx)
    tx_hash_bytes = w3.eth.send_raw_transaction(signed.rawTransaction)
    # web3.py returns HexBytes; .hex() is typed loosely in some versions,
    # so cast to str to keep `mypy --strict` happy.
    return str(tx_hash_bytes.hex())
